### PR TITLE
Remove async await from component code

### PR DESCRIPTION
**Pull request by vhfmag**
As discussed with @diegohaz [in the React Brasil Slack organization](https://react-brasil.slack.com/archives/C1EDVPBQC/p1558698085019100), using async in library code has issues, namely:

- If you ship it transpiled, it may cause websites that use it to have several polyfills for the same features
- It you ship it non-transpiled, you put in the users the burden of transpiling it. Transpilling dependencies is not necessarily a widespread practice, so the async code may end up in production code

Because of that, @diegohaz 	decided to drop async usage, and that's the aim of this PR.

Additionally, I installed the `eslint-plugin-es`  plugin to catch not-allowed async usage as lint errors.

**Does this PR introduce a breaking change?**

No :)

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -82,6 +82,22 @@
...
+        "**/packages/reakit-system-palette/**/*"
+      ],
+      "parserOptions": {
+        "ecmaVersion": 2018
+      },
+      "excludedFiles": ["**/__tests__/**/*"],
+      "rules": {
+        "es/no-async-functions": "error",
+        "es/no-async-iteration": "error"
+      }
```
Comment (diegohaz) [inline .eslintrc]: I think the added complexity doesn't worth the minimal gain it provides. We can spot the rare cases when async/await is used on code review.

File: packages/reakit/src/Form/FormState.ts [hunk 2]
```diff
@@ -306,23 +303,29 @@ export function unstable_useFormState<V = Record<any, any>>(
...
+            return response;
           }
-          const messages = await response;
-          dispatch({ type: "endValidate", messages });
+          return undefined;
+        })
+        .then(messages => {
+          if (messages) {
+            dispatch({ type: "endValidate", messages });
+          }
```
Comment (diegohaz) [inline packages/reakit/src/Form/FormState.ts:312]: It's possible that the `onValidate` function returns without messages. So I guess the check should be removed.

File: packages/reakit/src/Form/FormState.ts [hunk 3]
```diff
@@ -346,24 +349,34 @@ export function unstable_useFormState<V = Record<any, any>>(
...
-      } catch (errors) {
-        dispatch({ type: "endSubmit", errors });
-      }
-    }, [validate]),
+    submit: React.useCallback(
+      () =>
+        Promise.resolve()
+          .then(() => {
+            dispatch({ type: "startSubmit" });
+            return validate();
```
Comment (diegohaz) [inline packages/reakit/src/Form/FormState.ts]: Hmm, I think this dispatch should synchronous (outside the Promise).

Comment (diegohaz) [reply packages/reakit/src/Form/FormState.ts]: Also, `validate` is a promise. So I guess we can do `validate().then(...)`

File: packages/reakit/src/Form/__tests__/index-test.tsx [hunk 4]
```diff
@@ -37,13 +37,14 @@ test(
     const input = getByLabelText("input");
     expect(onValidate).not.toHaveBeenCalled();
     fireEvent.change(input, { target: { value: "a" } });
+    await Promise.resolve();
```
Comment (diegohaz) [inline packages/reakit/src/Form/__tests__/index-test.tsx]: Why is this necessary? Looks like validate on change doesn't work when testing on the browser.

File: .eslintrc (hunk 1 above)
Comment (vhfmag) [reply .eslintrc]: Okay, I'll revert that

File: packages/reakit/src/Form/FormState.ts (hunk 3 above)
Comment (vhfmag) [reply packages/reakit/src/Form/FormState.ts]: Both true. Fixed!

File: packages/reakit/src/Form/FormState.ts (hunk 2 above)
Comment (vhfmag) [reply packages/reakit/src/Form/FormState.ts:312]: It couldn't be removed because otherwise it would run not only after a `onValidate` call but also after the `return undefined`. I refactored it.

File: .eslintrc (hunk 1 above)
Comment (vhfmag) [reply .eslintrc]: Done

File: packages/reakit/src/Form/__tests__/index-test.tsx (hunk 4 above)
Comment (vhfmag) [reply packages/reakit/src/Form/__tests__/index-test.tsx]: Because of how I implemented `validate`, `onValidate` is only called after the main thread is free. I can refactor `Promise.resolve().then` into `new Promise(resolve =>` and this will be unnecessary, should I do it?

**Comment by vhfmag** [comment]
The proposed changes fixed the failing test I reported.

File: packages/reakit/src/Form/__tests__/index-test.tsx (hunk 4 above)
Comment (vhfmag) [reply packages/reakit/src/Form/__tests__/index-test.tsx]: I think the issue with it not working on the browser was caused by the `if (messages) { dispatch(/* ... */) }` you reported. I'll test it

Comment (vhfmag) [reply packages/reakit/src/Form/__tests__/index-test.tsx]: Tested it and it seems to be working, but if you prefer it to be sync, just tell me.

Comment (diegohaz) [reply packages/reakit/src/Form/__tests__/index-test.tsx]: Nice! What if we did `Promise.resolve(onValidate()).then(...)` instead? It must work on these scenarios:
- `onValidate` may be undefined;
- `onValidate` may be a normal function that returns undefined or an object of messages;
- `onValidate` may be a normal function that throws an object of errors;
- `onValidate` may be a promise that returns undefined of an object of messages;
- `onValidate` may be a promise that throws an object of errors.

Comment (vhfmag) [reply packages/reakit/src/Form/__tests__/index-test.tsx]: I did it like that:
```ts
const validate = React.useCallback(
  (vals = state.values) =>
    new Promise(resolve => {
      if (onValidate) {
        const response = onValidate(filterAllEmpty(vals));
        if (isPromise(response)) {
          dispatch({ type: "startValidate" });
        }

        resolve(
          Promise.resolve(response).then(messages => {
            dispatch({ type: "endValidate", messages });
            return messages;
          })
        );
      } else {
        resolve(undefined);
      }
    }).catch(errors => {
      dispatch({ type: "endValidate", errors });
      throw errors;
    }),
  [state.values, onValidate]
);
```

Comment (vhfmag) [reply packages/reakit/src/Form/__tests__/index-test.tsx]: As the function passed to `Promise`'s constructor is run synchronously, everything up to the `resolve` call is sync.

Comment (vhfmag) [reply packages/reakit/src/Form/__tests__/index-test.tsx]: I pushed it for you to see: https://github.com/reakit/reakit/pull/365/commits/4c283e1538002cdc84f94ee37e05ab9845cf1485

Comment (diegohaz) [reply packages/reakit/src/Form/__tests__/index-test.tsx]: Cool! Have a meeting now. Will take a look tonight. Thank you ❤️

**Comment by diegohaz** [comment]
Thanks a lot for contributing!

Based on our [community guidelines](https://github.com/reakit/reakit/blob/master/CONTRIBUTING.md), every person who has a PR of any kind merged is offered **an invitation to the Reakit organization**.

Should you accept, you'll get write access to the main repository and a fancy [Reakit logo](https://github.com/reakit/reakit/tree/master/logo) on your GitHub profile. You'll be able to label and close issues, create new branches etc. Make sure to read our [contribution and community guidelines](https://github.com/reakit/reakit/blob/master/CONTRIBUTING.md), which explains all of this in more detail.

If you have any questions, let me know!
